Runtime/Scripts/MobiledgeXAgent.cs

```diff
@@ -0,0 +1,162 @@
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using UnityEngine;
+
+namespace MobiledgeX
+{
+  // Unity Main Thread context call agent. TODO: Generalize to "Run on UI Thread".
+  public class MobiledgeXAgent : MonoBehaviour
+  {
+    Action<string> NotifyHashAvertisingIdCallers;
+
+    static private string HashedAdvertisingIdPrivate = null;
+    public string HashedAdvertisingId
+    {
+      get
+      {
+        string tmp = null;
+        lock (this)
+        {
+          if (MobiledgeXAgent.HashedAdvertisingIdPrivate != null)
+          {
+            tmp = String.Copy(MobiledgeXAgent.HashedAdvertisingIdPrivate);
+          }
+        }
+        return tmp;
+      }
+      set
+      {
+        lock (this)
+        {
+          MobiledgeXAgent.HashedAdvertisingIdPrivate = value;
+        }
+      }
+    }
+
+    public const int DEFAULT_ADVERTISING_ID_TIMEOUT_MS = 3000;
+    public int HashedAdvertisingIdTimeoutMs { get; set; } = DEFAULT_ADVERTISING_ID_TIMEOUT_MS;
+
+    private bool hashAdvertisingIdRequestedPrivate = false;
+    private bool hashAdvertisingIdRequested
+    {
+      get
+      {
+        string tmp = null;
+        lock (this)
+        {
+          return hashAdvertisingIdRequestedPrivate;
+        }
+      }
+      set
+      {
+        lock (this)
+        {
+          hashAdvertisingIdRequestedPrivate = value;
+        }
+      }
+    }
+
+    // Start is called before the first frame update
+    IEnumerator Start()
+    {
+      Debug.Log("MobiledgeXAgent: Created MobiledgeXAgent.");
+      yield return StartCoroutine(GetHashedAdvertisingID());
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+      if (hashAdvertisingIdRequested)
+      {
+        //Debug.Log("ZZZ: Starting AdID coroutine:");
+        StartCoroutine(GetHashedAdvertisingID());
```

Comment:
I don't know if you need to call **GetHashedAdvertisingID** in Update also
Although you have a lock its being called two times
<img width="711" alt="Screen Shot 2020-08-03 at 11 27 44 AM" src="https://user-images.githubusercontent.com/24863504/89214951-99ba3600-d57c-11ea-921b-3aed9b352284.png">

Reply:
Removed function.